[InstCombine] Improve inbounds preservation for ADD+GEP -> GEP+GEP

bjope opened the pull request:
Given that we have a "add nuw" and a "getelementptr inbounds nuw" like this:
   %idx = add nuw i64 %idx1, %idx2
   %gep = getelementptr inbounds nuw i32, ptr %ptr, i64 %idx

Then we can preserve the "inbounds nuw" flag when transforming that into two getelementptr instructions:
   %gep1 = getelementptr inbounds nuw i32, ptr %ptr, i64 %idx1
   %gep = getelementptr inbounds nuw i32, ptr %ptr, i64 %idx2

Similarly for just having "nuw" instead of "inbounds nuw" on the getelementptr.

Proof: https://alive2.llvm.org/ce/z/4uhfDq

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3087,12 +3087,22 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
-    // We can only preserve inbounds if the original gep is inbounds, the add
-    // is nsw, and the add operands are non-negative.
-    auto CanPreserveInBounds = [&](bool AddIsNSW, Value *Idx1, Value *Idx2) {
+    auto CanPreserveNoWrapFlags = [&](bool AddIsNSW, bool AddIsNUW, Value *Idx1,
```
Rename this to GetPreservedNoWrapFlags or something.

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3087,12 +3087,22 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
+                                      Value *Idx2) {
+      // Preserve "inbounds nuw" if the original gep is "inbounds nuw",
+      // and the add is "nuw".
+      if (GEP.isInBounds() && GEP.hasNoUnsignedWrap() && AddIsNUW)
+        return GEPNoWrapFlags::inBounds() | GEPNoWrapFlags::noUnsignedWrap();
```
```suggestion
      if (GEP.hasNoUnsignedWrap() && AddIsNUW)
        return GEP.getNoWrapFlags();
```
Would this work to subsume both this case and the only nuw one below?

bjope commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3087,12 +3087,22 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
-             isKnownNonNegative(Idx2, Q);
+      if (GEP.isInBounds() && AddIsNSW && isKnownNonNegative(Idx1, Q) &&
+          isKnownNonNegative(Idx2, Q))
+        return GEPNoWrapFlags::inBounds();
```
Right, we can probably skip it, but then we need to move the "nusw + nneg -> nuw" case (from line 3171) to make sure we infer nuw on the gep prior to this pattern.

bjope committed:
Fixups after review feedback:

- Make sure we try to derive inbound + nuw before doing the
  ADD+GEP->GEP+GEP rewrites. This to make it possible to rely
  on "nuw" being present when trying to preserve flags.

- Make sure the special ADDNSW+SEXTLIKE+GEP->GEP+GEP case is checking
  for ADDNUW+ZEXTNNEG+GEP->GEP+GEP when preserving flags.
  This way we do not need to check for non-negative operands.

bjope committed:
Fix clang-format

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3086,13 +3086,51 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
+      if (GEP.hasNoUnsignedWrap() && AddIsNUW) {
+        Flags |= GEPNoWrapFlags::noUnsignedWrap();
+        if (GEP.isInBounds())
+          Flags |= GEPNoWrapFlags::inBounds();
```
Same comment here, we should just be returning the original Flags instead of reconstructing them. Note that this works for all of `nuw`, `inbounds nuw` and `nusw nuw`: https://alive2.llvm.org/ce/z/QSweWW

bjope replied:
Changing to 16-bit ptr/index size in your example  seem to indicate that this isn't valid for just `nuw` (at least no unless we check that the types used for the offset satisfy the index size): https://alive2.llvm.org/ce/z/oyAv6T

My idea of reconstructing flags was to make the code more defensive. I have no clue what flags will be added in the future. If someone adds a "PointerIsEven" flag (or whatever) in the future, it might be wrong to preserve all flags. So I used a more defensive programming style to only preserve the flags that I explicitly handle. But maybe that isn't the LLVM style of programming? But maybe adding a helper in `GEPNoWrapFlags` (a reverse version of `intersectForOffsetAdd`) would be better then. Then one would need to consider updating that helper if adding/removing flags in `GEPNoWrapFlags`.

nikic replied:
> Changing to 16-bit ptr/index size in your example seem to indicate that this isn't valid for just `nuw` (at least no unless we check that the types used for the offset satisfy the index size): https://alive2.llvm.org/ce/z/oyAv6T

If you don't match the index size, you are introducing an implicit sext() on the indices. The transform is not valid in that case.

> My idea of reconstructing flags was to make the code more defensive. I have no clue what flags will be added in the future. If someone adds a "PointerIsEven" flag (or whatever) in the future, it might be wrong to preserve all flags. So I used a more defensive programming style to only preserve the flags that I explicitly handle. But maybe that isn't the LLVM style of programming? But maybe adding a helper in `GEPNoWrapFlags` (a reverse version of `intersectForOffsetAdd`) would be better then. Then one would need to consider updating that helper if adding/removing flags in `GEPNoWrapFlags`.

I don't think we should try too hard to predict the future and go with the simple solution here.

bjope replied:
> If you don't match the index size, you are introducing an implicit sext() on the indices. The transform is not valid in that case.
Oh, right, the transform isn't even valid if dropping the flags in that case.

It puzzeled me for awhile how implicit sext/trunc of indices would impact these rewrites (also for the legacy code with NSW considering that there are no checks on index size). Now I found out that we make sure that the gep index operand is using the correct index type size earlier in this function.  So we can expect that we match with the index size when doing the rewrites into GEP+GEP here.

nikic commented on llvm/test/Transforms/InstCombine/array.ll:
```diff
@@ -122,25 +122,23 @@ define ptr @gep_inbounds_nuwaddlike(ptr %ptr, i64 %a, i64 %b) {
-; CHECK-NEXT:    [[TMP1:%.*]] = getelementptr i32, ptr [[PTR]], i64 [[A]]
-; CHECK-NEXT:    [[GEP:%.*]] = getelementptr i32, ptr [[TMP1]], i64 [[B]]
+; CHECK-NEXT:    [[TMP1:%.*]] = getelementptr inbounds nuw i32, ptr [[PTR]], i64 [[A]]
+; CHECK-NEXT:    [[GEP:%.*]] = getelementptr inbounds nuw i32, ptr [[TMP1]], i64 [[B]]
```
Can you please also add a test for nusw nuw for completeness?

bjope replied:
Yes. Added @gep_inbounds_add_nusw_nuw.

bjope committed:
Fixup

bjope committed:
[InstCombine] Improve no-wrap flag preservation for ADD+GEP -> GEP+GEP

Given that we have a "add nuw" and a "getelementptr inbounds nuw" like
this:
   %idx = add nuw i64 %idx1, %idx2
   %gep = getelementptr inbounds nuw i32, ptr %ptr, i64 %idx

Then we can preserve the "inbounds nuw" flag when transforming that
into two getelementptr instructions:
   %gep1 = getelementptr inbounds nuw i32, ptr %ptr, i64 %idx1
   %gep = getelementptr inbounds nuw i32, ptr %ptr, i64 %idx2

Similarly for just having "nuw" or "nusw nuw" instead of "inbounds nuw"
on the getelementptr.

Proof: https://alive2.llvm.org/ce/z/QSweWW

bjope committed:
Fixup: Also handle "nusw nuw"

bjope commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3086,13 +3086,51 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
-      return GEP.isInBounds() && AddIsNSW && isKnownNonNegative(Idx1, Q) &&
-             isKnownNonNegative(Idx2, Q);
+    auto GetPreservedNoWrapFlags = [&](bool AddIsNUW, Value *Idx1,
+                                       Value *Idx2) {
```
Yes. Done.

nikic reviewed (approved):
LGTM, thanks!

bjope committed:
Fixup code comments

bjope committed:
Fixup formatting
